src/SpectatorView.Unity/Assets/SpectatorView/Scripts/Socketer/TCPConnectionManager.cs

```diff
@@ -3,61 +3,75 @@
     {
-        /// <summary>
-        /// Called when a client or server connection is established and the connection manager is using the TCP protocol.
-        /// </summary>
-        public event Action<SocketEndpoint> OnConnected;
-        /// <summary>
-        /// Called when a client or server connection is disconnected and the connection manager is using the TCP protocol.
-        /// </summary>
-        public event Action<SocketEndpoint> OnDisconnected;
-        /// <summary>
-        /// Called when a data payload is received
-        /// </summary>
-        public event Action<IncomingMessage> OnReceive;
         /// <summary>
         /// If true, socket clients will attempt to reconnect when disconnected.
         /// </summary>
-        public bool AttemptReconnectWhenClient = true;
+        [Tooltip("If true, socket clients will attempt to reconnect when disconnected.")]
+        [SerializeField]
+        public bool AttemptReconnectWhenClient = false;
+
+        /// <inheritdoc />
+        public event Action<INetworkConnection> OnConnected;
+
+        /// <inheritdoc />
+        public event Action<INetworkConnection> OnDisconnected;
+
+        /// <inheritdoc />
+        public event Action<IncomingMessage> OnReceive;
 
         private readonly TimeSpan timeoutInterval = TimeSpan.Zero;
-        private readonly ConcurrentQueue<SocketEndpoint> newConnections = new ConcurrentQueue<SocketEndpoint>();
-        private readonly ConcurrentQueue<SocketEndpoint> oldConnections = new ConcurrentQueue<SocketEndpoint>();
-        private readonly ConcurrentDictionary<int, SocketEndpoint> serverConnections = new ConcurrentDictionary<int, SocketEndpoint>();
+        private readonly ConcurrentQueue<TCPSocketEndpoint> newConnections = new ConcurrentQueue<TCPSocketEndpoint>();
+        private readonly ConcurrentQueue<TCPSocketEndpoint> oldConnections = new ConcurrentQueue<TCPSocketEndpoint>();
+        private readonly ConcurrentDictionary<int, TCPSocketEndpoint> serverConnections = new ConcurrentDictionary<int, TCPSocketEndpoint>();
         private readonly ConcurrentQueue<IncomingMessage> inputMessageQueue = new ConcurrentQueue<IncomingMessage>();
-        private SocketEndpoint clientConnection;
+        private TCPSocketEndpoint clientConnection;
         private SocketerClient client;
 
         private SocketerClient server;
 
-        /// <summary>
-        /// Returns true if any server or client connections exist, otherwise false
-        /// </summary>
+        /// <inheritdoc />
+        public IReadOnlyList<INetworkConnection> Connections
+        {
+            get
+            {
+                List<INetworkConnection> connections = new List<INetworkConnection>();
+                if (clientConnection != null)
+                {
+                    connections.Add(clientConnection);
+                }
+
+                if (serverConnections.Count > 0)
```

Comment:
`if (!serverConnections.IsEmpty())` please.

The internal implementation of `ConcurrentDictionary` [works like this](https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/Concurrent/ConcurrentDictionary.cs):

- `IsEmpty()` returns immediately as soon as it finds any elements in any of the tables.
- `Count` won't stop until it iterates over **all of them**, adding all results together, even if it was obvious after the first table that there are elements.

Fixing other 12 instances of "Count > 0" would also be great. Don't know how many of them will become faster, but IsEmpty is just more idiomatic anyway.

Reply:
it looks like this function exists for ConcurrentDictionary but doesnt exist for Dictionary. I will update ConcurrentDictionaries.